File: src/domain/board-v7.ts
import type {ProjectionCardData} from './projection';
export type Bands=Record<'50'|'80',[number,number]>;
export type TeamEvidence={quantile_dots?:{value:number;mass:number;probability_lo:number;probability_hi:number}[];expected:number;actual:number|null;intervals:Bands;error:number|null;hits:Record<'50'|'80',boolean|null>;pit:number|null};
export type GameEvidence={interval_kind?:string;spread_model?:string;outdoors?:boolean|null;qualified_lock:boolean;distribution_hash?:string;teams:Partial<Record<'away'|'home',TeamEvidence>>};
export type Summary={teams:number;mae:number|null;inside80:number;eligible80:number;coverage:Record<'margin'|'total',Record<'50'|'80',{hit:number;n:number;rate:number|null}>>;pit:number[]};
export type Ranked={game_id:string;team:string;expected:number;actual:number;error:number;contributions:ProjectionCardData['contributions']['away']};
export type BoardEvidence={schema:string;board_sha256:string;games:Record<string,GameEvidence>;trust:Summary;weeks:{week:number;scope:string;engine:Summary;ours:Summary}[];closest:Ranked[];furthest:Ranked[];prior_seasons:{season:number;week:number;mae:number}[];reference:{oof_mae:number|null;climatology_mae:number|null;floor:number|null;floor_status:string};edits:{engine:Summary;ours:Summary;best_tags:{game_id:string;mae:number;tags:string[]}[];worst_tags:{game_id:string;mae:number;tags:string[]}[]}};
export type Lens='POINTS'|'ERROR';
export const integer=(n:number|null|undefined)=>n==null||!Number.isFinite(n)?'—':String(Math.sign(n)*Math.floor(Math.abs(n)+.5));
export const signed=(n:number)=>n>0?'+'+integer(n):integer(n);
export const position=(v:number,lens:Lens)=>(Math.max(lens==='POINTS'?0:-30,Math.min(lens==='POINTS'?45:30,v))-(lens==='POINTS'?0:-30))/(lens==='POINTS'?45:60)*100;
export function eligible(g:ProjectionCardData,e?:GameEvidence){return g.status==='FINAL'&&!!g.final&&!!e?.qualified_lock;}
export function emptyLabel(g:ProjectionCardData,e?:GameEvidence){return g.status==='MISSED'||g.evidence==='RETROSPECTIVE'||g.status==='FINAL'&&!e?.qualified_lock?'no lock':'not played';}
export function sortGames(games:ProjectionCardData[],e:Record<string,GameEvidence>,sort:string){const kickoff=(a:ProjectionCardData,b:ProjectionCardData)=>Date.parse(a.kickoff_at)-Date.parse(b.kickoff_at)||a.game_id.localeCompare(b.game_id);return [...games].sort((a,b)=>{if(sort==='kickoff')return kickoff(a,b);const aa=eligible(a,e[a.game_id]),bb=eligible(b,e[b.game_id]);if(aa!==bb)return aa?-1:1;if(!aa)return kickoff(a,b);const error=(g:ProjectionCardData)=>Math.max(...Object.values(e[g.game_id].teams).map(t=>Math.abs(t!.error??0)));return (sort==='descending'?-1:1)*(error(a)-error(b))||kickoff(a,b);});}
export function status(e?:GameEvidence){const t=Object.values(e?.teams??{});return !t.length||t.some(x=>x?.actual==null)?'interval unavailable':t.every(x=>x?.hits['50'])?'in 50%':t.every(x=>x?.hits['80'])?'in 80%':'outside 80%';}
export const text=(s:string)=>s.replace(/(-?\d+)\.\d+/g,x=>integer(Number(x)));

// Open the active/upcoming published slate, not a stale publication default.
// Keep a slate through its final scheduled game window, then advance.
export function currentBoardWeek(board: {default_week:number;games:{week:number;kickoff_at:string}[]}, now=Date.now()):number {
 const ends=new Map<number,number>();
 for(const game of board.games){const kickoff=Date.parse(game.kickoff_at);if(Number.isFinite(kickoff))ends.set(game.week,Math.max(ends.get(game.week)??-Infinity,kickoff+4*60*60*1000));}
 const weeks=[...ends].sort((a,b)=>a[0]-b[0]);
 return weeks.find(([,end])=>end>now)?.[0]??weeks.at(-1)?.[0]??board.default_week;
}

File: engine/board_v7.py
"""Read-only presentation evidence. Never writes a forecast, lock, grade or fit."""
import statistics
from collections import defaultdict
from engine.projection.distribution import pmf,quantile

def qualified(g):
 return bool(g.get('projection') and g.get('freeze_time') and g.get('evidence')=='AS_ISSUED' and g.get('status') in ('LOCKED','FINAL'))

def metadata(g,shapes):
 if not g.get('projection'):return {'qualified_lock':False,'teams':{}}
 roof=(g.get('forecast') or {}).get('roof') or g.get('learning_features',{}).get('home',{}).get('game',{}).get('roof')
 out={'outdoors':None if roof is None else str(roof).lower() in ('outdoors','open'),'qualified_lock':qualified(g),'distribution_hash':shapes['team_points']['sha256'],'interval_kind':'single-game predictive','spread_model':'constant_within_issuing_version','teams':{}}
 for side in ('away','home'):
  expected=g['projection'][side+'_points'];mass=pmf(shapes['team_points'],expected)
  bands={str(level):[quantile(mass,(1-level/100)/2),quantile(mass,1-(1-level/100)/2)] for level in (50,80)}
  actual=g.get('final',{}).get(side+'_points') if qualified(g) else None
  out['teams'][side]={'quantile_dots':[{'value':quantile(mass,(i+.5)/10),'mass':.1,'probability_lo':i/10,'probability_hi':(i+1)/10} for i in range(10)],'expected':expected,'actual':actual,'intervals':bands,'error':None if actual is None else actual-expected,'hits':{k:None if actual is None else lo<=actual<=hi for k,(lo,hi) in bands.items()},'pit':None if actual is None else sum(p for x,p in mass.items() if x<actual)+sum(p for x,p in mass.items() if x==actual)/2}
 return out

def summary(games,meta,source='PROJECTION'):
 rows=[];hits={t:{l:[] for l in ('50','80')} for t in ('margin','total')};teamhits=[];pit=[0]*10
 for g in games:
  if not meta[g['game_id']]['qualified_lock'] or not g.get('final'):continue
  pred=g.get('projection' if source=='PROJECTION' else 'ours')
  if not pred:continue
  actual=g['final']; errors=[actual[s+'_points']-pred[s+'_points'] for s in ('away','home')];rows+=errors
  vals={'margin':actual['home_points']-actual['away_points'],'total':actual['home_points']+actual['away_points']}
  for t in hits:
   for l in hits[t]:
    band=pred.get('intervals',{}).get(t,{}).get(l)
    if band:hits[t][l].append(band[0]<=vals[t]<=band[1])
  if source=='PROJECTION':
   for team in meta[g['game_id']]['teams'].values():
    teamhits.append(team['hits']['80']);pit[min(9,int(team['pit']*10))]+=1
 return {'teams':len(rows),'mae':statistics.mean(map(abs,rows)) if rows else None,'inside80':sum(teamhits),'eligible80':len(teamhits),'coverage':{t:{l:{'hit':sum(v),'n':len(v),'rate':statistics.mean(v) if v else None} for l,v in levels.items()} for t,levels in hits.items()},'pit':pit}

def build(board,shape_loader,historical,climatology_training=()):
 games=board['games'];meta={g['game_id']:metadata(g,shape_loader(g)) if g.get('projection') else {'qualified_lock':False,'teams':{}} for g in games}
 weeks=[]
 for week in sorted({g['week'] for g in games}):
  for scope in ('week','cumulative'):
   subset=[g for g in games if g['week']==week or scope=='cumulative' and g['week']<=week]
   weeks.append({'week':week,'scope':scope,'engine':summary(subset,meta),'ours':summary(subset,meta,'OURS')})
 ranked=[]
 for g in games:
  if not meta[g['game_id']]['qualified_lock'] or not g.get('final'):continue
  for side,t in meta[g['game_id']]['teams'].items():ranked.append({'game_id':g['game_id'],'team':g[side],'expected':t['expected'],'actual':t['actual'],'error':t['error'],'contributions':g.get('contributions',{}).get(side,[])})
 ranked.sort(key=lambda x:(abs(x['error']),x['game_id'],x['team']))
 historical_weeks=defaultdict(list)
 for r in historical:
  historical_weeks[(r['season'],r['week'])].extend([abs(r['actual_home']-r['home']),abs(r['actual_away']-r['away'])])
 prior=[{'season':s,'week':w,'mae':statistics.mean(v)} for (s,w),v in sorted(historical_weeks.items()) if board['games'][0]['season']-5<=s<board['games'][0]['season']]
 allerrors=[v for values in historical_weeks.values() for v in values]
 # Expanding earlier-season league mean is the climatology reference, never all-history future data.
 baseline=[]
 means={season:statistics.mean([p['actual_points'] for p in climatology_training if p['season']<season and p.get('actual_points') is not None]) for season in {r['season'] for r in historical} if any(p['season']<season and p.get('actual_points') is not None for p in climatology_training)}
 for r in historical:
  if r['season'] in means:baseline.extend(abs(r[k]-means[r['season']]) for k in ('actual_home','actual_away'))

 return {'schema':'board-v7-evidence','board_sha256':board['content_sha256'],'published_at':board['published_at'],'games':meta,'trust':summary(games,meta),'weeks':weeks,'closest':ranked[:5],'furthest':list(reversed(ranked[-5:])),'prior_seasons':prior,'reference':{'oof_mae':statistics.mean(allerrors) if allerrors else None,'climatology_mae':statistics.mean(baseline) if baseline else None,'floor':None,'floor_status':'Uncomputed — governed E4 experiment'},'edits':{'engine':summary([g for g in games if g.get('ours')],meta),'ours':summary(games,meta,'OURS'),'best_tags':[],'worst_tags':[]}}
